### PR TITLE
tests/gasnet: convert to new stand-alone test process

### PR DESCRIPTION
**Pull request by tldahlgren**
Converted stand-alone tests to use the new test process.

```
$ spack -v test run gasnet
==> Spack test 4f4qi3veyt4w3sdo5xqc2gcocunzkpdg
==> Testing package gasnet-2023.3.0-ihw5qpj
==> [2023-05-22-12:34:34.408695] test: test_testgasnet: run testgasnet and check output
==> [2023-05-22-12:34:34.409367] test: test_testgasnet_smp: run smp-conduit/testgasnet
==> [2023-05-22-12:34:34.411552] '/usr/bin/env' 'GASNET_PSHM_NODES=4' '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/smp/testgasnet'
=====> testgasnet nprocs=4 config=RELEASE=2023.3.0,SPEC=0.16,CONDUIT=SMP(SMP-2.0/SMP-2.0),THREADMODEL=PAR,SEGMENT=FAST,PTR=64bit,CACHE_LINE_BYTES=64,noalign,pshm,nodebug,notrace,nostats,nodebugmalloc,nosrclines,timers_native,membars_native,atomics_native,atomic32_native,atomic64_native,notiopt compiler=GNU/10.3.1 sys=x86_64-pc-linux-gnu
node 0/4 hostname is: quartz2306 (supernode=0 pid=3053362)
node 3/4 hostname is: quartz2306 (supernode=0 pid=3053365)
node 2/4 hostname is: quartz2306 (supernode=0 pid=3053364)
node 1/4 hostname is: quartz2306 (supernode=0 pid=3053363)
node 0/4 testgasnet running...
SMP conduit: v2.0 GASNET_ALIGNED_SEGMENTS=0
 gex_AM_MaxArgs():             16
 gex_AM_LUBRequestMedium(): 65416
 gex_AM_LUBReplyMedium():   65416
 gex_AM_LUBRequestLong():   2147483647
 gex_AM_LUBReplyLong():     2147483647
 MaxLocalSegmentSize on node0:  15 GB
 MaxGlobalSegmentSize:          15 GB
my args: argc=1 argv=['/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/smp/testgasnet']
my args: argc=1 argv=['/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/smp/testgasnet']
my args: argc=1 argv=['/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/smp/testgasnet']
my args: argc=1 argv=['/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/smp/testgasnet']
[2] Invoking USER for backtrace...
[0] Invoking USER for backtrace...
[1] Invoking USER for backtrace...
[3] Invoking USER for backtrace...
node 1/4 *** passed libgasnet_tools test!!
node 3/4 *** passed libgasnet_tools test!!
node 0/4 *** passed libgasnet_tools test!!
node 2/4 *** passed libgasnet_tools test!!
node 3/4 *** passed object test!!
node 2/4 *** passed object test!!
node 1/4 *** passed object test!!
node 0/4 *** passed object test!!
node 3/4 *** passed blocking test!!
node 0/4 *** passed blocking test!!
node 2/4 *** passed blocking test!!
node 1/4 *** passed blocking test!!
node 1/4 *** passed blocking list test!!
node 0/4 *** passed blocking list test!!
node 3/4 *** passed blocking list test!!
node 2/4 *** passed blocking list test!!
node 3/4 *** passed nbi test!!
node 2/4 *** passed nbi test!!
node 0/4 *** passed nbi test!!
node 1/4 *** passed nbi test!!
node 3/4 *** passed value test!!
node 2/4 *** passed value test!!
node 0/4 *** passed value test!!
node 1/4 *** passed value test!!
node 1/4 *** passed nb put/overwrite test!!
node 2/4 *** passed nb put/overwrite test!!
node 0/4 *** passed nb put/overwrite test!!
node 3/4 *** passed nb put/overwrite test!!
node 3/4 *** passed nbi put/overwrite test!!
node 1/4 *** passed nbi put/overwrite test!!
node 2/4 *** passed nbi put/overwrite test!!
node 0/4 *** passed nbi put/overwrite test!!
node 1/4 *** passed AM test!!
node 0/4 *** passed AM test!!
node 2/4 *** passed AM test!!
node 3/4 *** passed AM test!!
node 3/4 done.
node 0/4 done.
node 1/4 done.
node 2/4 done.
*** gasnetc_init (quartz2306:3053362): about to spawn...
*** gasnetc_init (proc 1): spawn successful - proc 1/4 starting...
ENV parameter: GASNET_TSC_VERBOSE = NO                          (default)
ENV parameter: GASNET_TSC_RATE = wallclock                      (default)
ENV parameter: GASNET_TSC_RATE_TOLERANCE = 0.0005               (default)
ENV parameter: GASNET_TSC_RATE_HARD_TOLERANCE = 0.02            (default)
ENV parameter: GASNET_SPAWN_VERBOSE = YES                                
ENV parameter: GASNET_PSHM_NODES = 4                                     
ENV parameter: GASNET_EXITTIMEOUT_MAX = 5                       (default)
ENV parameter: GASNET_EXITTIMEOUT_MIN = 1                       (default)
ENV parameter: GASNET_EXITTIMEOUT_FACTOR = 0.1                  (default)
ENV parameter: GASNET_EXITTIMEOUT = 1.4                         (default)
ENV parameter: GASNET_FREEZE = NO                               (default)
ENV parameter: GASNET_DISABLE_ARGDECODE = NO                    (default)
ENV parameter: GASNET_BACKTRACE = NO                            (default)
ENV parameter: GASNET_BACKTRACE_MT = YES                        (default)
ENV parameter: GASNET_TMPDIR = *not set*                        (default)
ENV parameter: GASNET_DISABLE_ENVDECODE = NO                    (default)
ENV parameter: TMPDIR = /var/tmp/dahlgren                                
ENV parameter: GASNET_BACKTRACE_TYPE = GDB,LLDB,GSTACK,PSTACK,EXECINFO,USER   (default)
*** gasnetc_init (proc 2): spawn successful - proc 2/4 starting...
ENV parameter: GASNET_FREEZE_ON_ERROR = NO                      (default)
ENV parameter: GASNET_FREEZE_SIGNAL = *not set*                 (default)
ENV parameter: GASNET_BACKTRACE_SIGNAL = *not set*              (default)
ENV parameter: GASNET_SUPERNODE_MAXSIZE = 0                              
*** gasnetc_init (proc 0): spawn successful - proc 0/4 starting...
ENV parameter: GASNET_PSHM_NETWORK_DEPTH = 32                   (default)
ENV parameter: GASNET_AMPSHM_MEMORY_REPORT = NO                 (default)
*** gasnetc_init (proc 3): spawn successful - proc 3/4 starting...
ENV parameter: GASNET_BARRIER = PSHM                            (default)
ENV parameter: GASNET_COLL_SCRATCH_SIZE = 2 MB                  (default)
ENV parameter: GASNET_MAX_SEGSIZE = 26 GB                       (default)
ENV parameter: GASNET_CATCH_EXIT = YES                          (default)
ENV parameter: GASNET_DISABLE_MUNMAP = NO                       (default)
ENV parameter: GASNET_FS_SYNC = NO                              (default)
ENV parameter: GASNET_MAX_THREADS = 256                         (default)
ENV parameter: GASNET_PSHM_BARRIER_RADIX = 0                    (default)
ENV parameter: GASNET_COLL_P2P_EAGER_MIN = 16                   (default)
ENV parameter: GASNET_COLL_P2P_EAGER_SCALE = 16                 (default)
ENV parameter: GASNET_COLL_ROOTED_GEOM = KNOMIAL_TREE,2         (default)
ENV parameter: GASNET_COLL_BROADCAST_GEOM = KNOMIAL_TREE,2      (default)
ENV parameter: GASNET_COLL_SCATTER_GEOM = KNOMIAL_TREE,2        (default)
ENV parameter: GASNET_COLL_GATHER_GEOM = KNOMIAL_TREE,2         (default)
ENV parameter: GASNET_COLL_GATHER_ALL_DISSEM_LIMIT_PER_THREAD = 1 KB   (default)
ENV parameter: GASNET_COLL_GATHER_ALL_DISSEM_LIMIT = 1 KB       (default)
ENV parameter: GASNET_COLL_EXCHANGE_DISSEM_LIMIT_PER_THREAD = 1 KB   (default)
ENV parameter: GASNET_COLL_EXCHANGE_DISSEM_LIMIT = 1 KB         (default)
ENV parameter: GASNET_COLL_EXCHANGE_DISSEM_RADIX = 2            (default)
ENV parameter: GASNET_COLL_PIPE_SEG_SIZE = 512 KB               (default)
ENV parameter: GASNET_COLL_AUTOTUNE_WARM_ITERS = 5              (default)
ENV parameter: GASNET_COLL_AUTOTUNE_PERF_ITERS = 10             (default)
ENV parameter: GASNET_COLL_AUTOTUNE_ALLOW_FLAT_TREE = 1         (default)
ENV parameter: GASNET_COLL_TUNING_FILE = *not set*              (default)
ENV parameter: GASNET_COLL_PRINT_AUTOTUNE_TIMER = NO            (default)
ENV parameter: GASNET_COLL_PRINT_COLL_ALG = NO                  (default)
ENV parameter: GASNET_COLL_ENABLE_SEARCH = NO                   (default)
ENV parameter: GASNET_COLL_ENABLE_PROFILE = NO                  (default)
ENV parameter: GASNET_COLL_REDUCE_GEOM = KNOMIAL_TREE,2         (default)
ENV parameter: GASNET_NO_SCRATCH_WARN_THRESHOLD = 8             (default)
ENV parameter: GASNET_TEST_POLITE_SYNC = NO                     (default)
ENV parameter: GASNET_TEST_THREAD_LIMIT = 1                     (default)
ENV parameter: GASNET_TEST_SET_AFFINITY = YES                   (default)
*** EXIT STATE (proc 2): gasnet_exit(0)
*** EXIT STATE (proc 3): gasnet_exit(0)
*** EXIT STATE (proc 1): gasnet_exit(0)
*** EXIT STATE (proc 0): gasnet_exit(0)
PASSED: Gasnet::test_testgasnet_smp
PASSED: Gasnet::test_testgasnet
==> [2023-05-22-12:34:34.534717] test: test_testtools: run testtools and check output
==> [2023-05-22-12:34:34.536420] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/testtools'
=====> testtools config=RELEASE=2023.3.0,SPEC=1.20,PTR=64bit,nodebug,PAR,timers_native,membars_native,atomics_native,atomic32_native,atomic64_native compiler=GNU/10.3.1 sys=x86_64-pc-linux-gnu
hostname is: quartz2306 (pid=3053371)
testtools running...
gasnett_exe_name()='/usr/WS1/dahlgren/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/gasnet-2023.3.0-ihw5qpjc6r4yg73dspgyr7e6tfgz3y27/tests/testtools'
Running testtools with 100 iterations and 10 threads
System page size is 2^12 == 4096
CPU count estimated to be: 36
Cache line size estimated to be: 64
Physical memory size estimated to be: 125 GB
A: Testing high-performance timers and sleep...
B: Testing zero-byte counting...
C: Testing local membars...
D: Testing local write membars...
E: Testing local read membars...
F: Testing threadkey (sequential)...
G: Testing atomic ops (sequential)...
H: Testing client-provided backtrace code...
Invoking USER for backtrace...
Spawning pthreads...
I: parallel atomic-op barrier test...
J: parallel threadkey test...
K: parallel atomic-op pounding test...
L: parallel dec-test pounding test...
M: parallel word-tearing test...
N: parallel membar test...
O: parallel compare-and-swap test...
P: parallel swap test...
Q: parallel add test...
R: parallel atomic-op fence test...
Done.
ENV parameter: GASNET_BACKTRACE = NO                            (default)
ENV parameter: GASNET_BACKTRACE_MT = YES                        (default)
ENV parameter: GASNET_TMPDIR = *not set*                        (default)
ENV parameter: TMPDIR = /var/tmp/dahlgren                                
ENV parameter: GASNET_BACKTRACE_TYPE = USER                              
ENV parameter: GASNET_FREEZE_ON_ERROR = NO                      (default)
ENV parameter: GASNET_FREEZE_SIGNAL = *not set*                 (default)
ENV parameter: GASNET_BACKTRACE_SIGNAL = *not set*              (default)
ENV parameter: GASNET_TEST_THREAD_LIMIT = 255                   (default)
ENV parameter: GASNET_FS_SYNC = NO                              (default)
ENV parameter: GASNET_MAXIMIZE_RLIMIT_CPU = YES                 (default)
ENV parameter: GASNET_MAXIMIZE_RLIMIT_DATA = YES                (default)
ENV parameter: GASNET_MAXIMIZE_RLIMIT_RSS = YES                 (default)
ENV parameter: GASNET_MAXIMIZE_RLIMIT_STACK = YES               (default)
ENV parameter: GASNET_MAXIMIZE_RLIMIT_AS = YES                  (default)
ENV parameter: GASNET_TEST_TIME_SLACK = 0.01                    (default)
ENV parameter: GASNET_TSC_VERBOSE = NO                          (default)
ENV parameter: GASNET_TSC_RATE = wallclock                      (default)
ENV parameter: GASNET_TSC_RATE_TOLERANCE = 0.0005               (default)
ENV parameter: GASNET_TSC_RATE_HARD_TOLERANCE = 0.02            (default)
PASSED: Gasnet::test_testtools
==> [2023-05-22-12:34:56.328640] Completed testing
==> [2023-05-22-12:34:56.328947] 
======================= SUMMARY: gasnet-2023.3.0-ihw5qpj =======================
Gasnet::test_testgasnet_smp .. PASSED
Gasnet::test_testgasnet .. PASSED
Gasnet::test_testtools .. PASSED
============================= 3 passed of 3 parts ==============================
============================== 1 passed of 1 spec ==============================
```

### CODE REVIEW

**Comment by bonachea** [comment]
I'm interested in testing and reviewing this PR, but might not have time in the next few weeks due to some release-related deadlines this month. The fact that it also depends on a PR not yet merged to develop further complicates matters.

@tldahlgren some questions:
1. What is the envisioned schedule for merging dependent PR #34236?
2. Does this package-specific PR need to be merged "together" with that wider PR (ie does the wider PR break the existing package testing DSL), or can it trail the merge of that wider PR by a few weeks?

**Comment by tldahlgren** [comment]
> I'm interested in testing and reviewing this PR, but might not have time in the next few weeks due to some release-related deadlines this month. The fact that it also depends on a PR not yet merged to develop further complicates matters.

Np.

> 
> @tldahlgren some questions:
> 
> 1. What is the envisioned schedule for merging dependent PR [Stand-alone testing: Make recipe support and processing pytest-like #34236](https://github.com/spack/spack/pull/34236)?

I'm trying to get #34236 ready for review this week and the hope is it is reviewed and merged relatively quickly once that is done.  

The PR has over 100 commits involving 100 files so the intention is to pair that PR down to the changes to the core to make it easier to review.

> 2. Does this package-specific PR need to be merged "together" with that wider PR (ie does the wider PR break the existing package testing DSL), or can it trail the merge of that wider PR by a few weeks?

The latter.  That is, #34236 will continue to support but deprecate `test()` methods until we can get all of the packages converted and tested.

**Comment by bonachea** [comment]
Hi @tldahlgren - 

I saw the new test support was added to develop (congrats!)
Thanks for rebasing this PR - I plan to carefully test it and review this week.

-D

**Comment by tldahlgren** [comment]
> Hi @tldahlgren -
> 
> I saw the new test support was added to develop (congrats!) Thanks for rebasing this PR - I plan to carefully test it and review this week.
> 
> -D

Thanks Dan.

**Comment by bonachea** [comment]
@tldahlgren All I've done so far is light testing, but I've already noticed that when I run `spack -v test run gasnet` a **second** time, I consistently get a trailing failure like this:

```
...
PASSED: Gasnet::test_testtools
==> [2023-05-22-15:40:30.786304] Completed testing
==> [2023-05-22-15:40:30.786433] 
======================= SUMMARY: gasnet-2023.3.0-hpbkx3u =======================
Gasnet::test_testgasnet_smp .. PASSED
Gasnet::test_testgasnet .. PASSED
Gasnet::test_testtools .. PASSED
============================= 3 passed of 3 parts ==============================
============================== 1 failed of 1 spec ==============================

See gasnet-2023.3.0-hpbkx3u test results at:
  <redacted>/gasnet-2023.3.0-hpbkx3u-test-out.txt
==> Error: 1 test(s) in the suite failed.
```
Where the named file ends with this output:
```
...
ENV parameter: GASNET_TSC_RATE_HARD_TOLERANCE = 0.02            (default)
PASSED: Gasnet::test_testtools
==> [2023-05-22-15:41:22.171967] Completed testing
==> [2023-05-22-15:41:22.172102] 
======================= SUMMARY: gasnet-2023.3.0-hpbkx3u =======================
Gasnet::test_testgasnet_smp .. PASSED
Gasnet::test_testgasnet .. PASSED
Gasnet::test_testtools .. PASSED
============================= 3 passed of 3 parts ==============================
Test failure: invalid literal for int() with base 10: '2\n2\n2'
```

That final line appears to be coming from somewhere outside the package and causing the whole test suite to fail.
Any idea what that's about?

**Comment by tldahlgren** [comment]
> @tldahlgren All I've done so far is light testing, but I've already noticed that when I run `spack -v test run gasnet` a **second** time, I consistently get a trailing failure like this:
> 
> ```
> ...
> PASSED: Gasnet::test_testtools
> ==> [2023-05-22-15:40:30.786304] Completed testing
> ==> [2023-05-22-15:40:30.786433] 
> ======================= SUMMARY: gasnet-2023.3.0-hpbkx3u =======================
> Gasnet::test_testgasnet_smp .. PASSED
> Gasnet::test_testgasnet .. PASSED
> Gasnet::test_testtools .. PASSED
> ============================= 3 passed of 3 parts ==============================
> ============================== 1 failed of 1 spec ==============================
> 
> See gasnet-2023.3.0-hpbkx3u test results at:
>   <redacted>/gasnet-2023.3.0-hpbkx3u-test-out.txt
> ==> Error: 1 test(s) in the suite failed.
> ```
> 
> Where the named file ends with this output:
> 
> ```
> ...
> ENV parameter: GASNET_TSC_RATE_HARD_TOLERANCE = 0.02            (default)
> PASSED: Gasnet::test_testtools
> ==> [2023-05-22-15:41:22.171967] Completed testing
> ==> [2023-05-22-15:41:22.172102] 
> ======================= SUMMARY: gasnet-2023.3.0-hpbkx3u =======================
> Gasnet::test_testgasnet_smp .. PASSED
> Gasnet::test_testgasnet .. PASSED
> Gasnet::test_testtools .. PASSED
> ============================= 3 passed of 3 parts ==============================
> Test failure: invalid literal for int() with base 10: '2\n2\n2'
> ```
> 
> That final line appears to be coming from somewhere outside the package and causing the whole test suite to fail. Any idea what that's about?

Sounds like a bug and I think I've heard something similar before.

I always run `spack test remove -y` between test runs so results aren't appended.  Let me try to reproduce this.

Thanks!

**Comment by tldahlgren** [comment]
> > ============================= 3 passed of 3 parts ==============================
> > Test failure: invalid literal for int() with base 10: '2\n2\n2'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That final line appears to be coming from somewhere outside the package and causing the whole test suite to fail. Any idea what that's about?
> 
> Sounds like a bug and I think I've heard something similar before.
> 
> I always run `spack test remove -y` between test runs so results aren't appended. Let me try to reproduce this.
> 
> Thanks!

From the error output string I'd say you ran it three times.  I've got a fix that I'm testing.

**Comment by bonachea** [comment]
> 
> From the error output string I'd say you ran it three times. I've got a fix that I'm testing.

You are probably correct, I was running simultaneously on multiple different systems and might have lost track. Another possibly "bad output" line I also saw:
```
Test failure: invalid literal for int() with base 10: '2\n2'
```
So I'm guessing that corresponds to the second run?

**Comment by tldahlgren** [comment]
> > From the error output string I'd say you ran it three times. I've got a fix that I'm testing.
> 
> You are probably correct, I was running simultaneously on multiple different systems and might have lost track. Another possibly "bad output" line I also saw:
> 
> ```
> Test failure: invalid literal for int() with base 10: '2\n2'
> ```
> 
> So I'm guessing that corresponds to the second run?

Yes.  The fix is in #37841 .

File: var/spack/repos/builtin/packages/gasnet/package.py [hunk 1]
```diff
@@ -157,11 +157,34 @@ def test_install(self):
                 purpose="Running testtools",
```
Comment (bonachea) [inline var/spack/repos/builtin/packages/gasnet/package.py]: This `self.run_test()` invocation now generates a warning:
```
==> [2023-05-22-15:15:05.630432] Warning: The 'run_test' method is deprecated. Use 'test_part' instead for gasnet to process testtools.
```
I think the fix is to replace this call with `self.test_testtools()`

File: var/spack/repos/builtin/packages/gasnet/package.py [hunk 2]
```diff
@@ -157,11 +157,34 @@ def test_install(self):
...
             )
 
-    def test(self):
+    def _setup_test_env(self):
+        """Set up key stand-alone test environment variables."""
+        # TBD: Are these relevant for both test executables?
+
+        os.environ["GASNET_VERBOSEENV"] = "1"  # include diagnostic info
```
Comment (bonachea) [inline var/spack/repos/builtin/packages/gasnet/package.py]: The answer to the question asked by this comment:
1. `GASNET_VERBOSEENV` is relevant to every GASNet test.
2. `GASNET_SPAWN_VERBOSE` and `GASNET_SSH_SERVERS` are irrelevant to testtools, but they are also 100% harmless.

I think it's fine to remove the comment and leave the code as-is:


```suggestion
        """Set up key stand-alone test environment variables."""
        os.environ["GASNET_VERBOSEENV"] = "1"  # include diagnostic info
```

**Review by bonachea — CHANGES_REQUESTED** [review]
Overall looks like a faithful conversion from the old stand-alone tests to the new infrastructure, and passes my light testing (aside from issue #37840). 

I made a few minor requests for improvement, but otherwise LGTM

File: var/spack/repos/builtin/packages/gasnet/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/gasnet/package.py]: Great catch!  Thanks.  I need to make sure I check build-time tests and replace those calls there as well.

**Review by bonachea — APPROVED** [review]
LGTM! 
Thanks for the fixes @tldahlgren

**Review by PHHargrove — APPROVED** [review]
LGTM

**Review by alalazo — APPROVED** [review]

**Pull request merged**